keps/sig-windows/20190714-windows-csi-support.md

```diff
@@ -302,15 +308,21 @@ Once compiled for Windows, container images based on Window Base images (like Na
 
-![Kubernetes CSI Node Components and Interactions](https://ddebroywin1.s3-us-west-2.amazonaws.com/csi-proxy2.png "Kubernetes CSI Node Components and Interactions")
+![Kubernetes CSI Node Components and Interactions](csi-proxy3.png?raw=true "Kubernetes CSI Node Components and Interactions")
 
 A GRPC based interface, CSIProxyService, will be used by CSI Node Plugins to invoke privileged operations on the host through csi-proxy.exe. CSIProxyService will be versioned and any release of csi-proxy.exe binary will strive to maintain backward compatibility across as many prior stable versions of the API as possible. This avoids having to run multiple versions of the csi-proxy.exe binary on the same Windows host if multiple CSI Node Plugins (that do not depend on APIs in Alpha or Beta versions) have been configured and the version of the csi-proxy API required by the plugins are different. For every version of the API supported, csi-proxy.exe will first probe for and then expose a `\\.\pipe\csi-proxy-v[N]` pipe where v[N] can be v1, v2alpha1, v3beta1, etc. If during the initial probe phase, csi-proxy.exe determines that another process is already listening on a `\\.\pipe\csi-proxy-v[N]` named pipe, it will not try to create and listen on that named pipe. This allows multiple versions of csi-proxy.exe to run side-by-side if absolutely necessary to support multiple CSI Node Plugins that require widely different versions of CSIProxyService that no single version of csi-proxy.exe can support.
 
-If an old stable version of CSIProxyService, say vN, can no longer be supported, support for it may be dropped from the latest version of csi-proxy.exe. To continue running CSI Node Plugins that depend on the old version vN, Kubernetes administrators will be required to run the latest version of the csi-proxy.exe (that will be used by CSI Node Plugins that use versions of CSIProxyService more recent than vN) along with an old version of csi-proxy.exe that does support vN.
+##### CSI Proxy Configuration
+There will be two command line parameters that may be passed to csi-proxy.exe:
+1. kubernetes-csi-plugins-path: String parameter pointing to the path used by Kubernetes CSI plugins. Will default to: `C:\var\lib\kubelet\plugins\kubernetes.io\csi`. All requests for creation and deletion of paths through the CSIProxyService RPCs (detailed below) will need to be under this path.
```

Comment:
> will need to be under this path

Including resolution of any links in the parent dir?

For example, in the following setup:
1. `kubernetes-csi-plugins-path` = `C:\var\lib\kubelet\plugins\kubernetes.io\csi`
2. `C:\var\lib\kubelet\plugins\kubernetes.io\csi\foo\bar` is a link that points to `Z:\baz`

Would a request to delete `C:\var\lib\kubelet\plugins\kubernetes.io\csi\foo\bar\quux` be allowed (since it really resolves to `Z:\baz\quux`?

Reply:
@msau42 is more familiar with what the CSI drivers are required to do, and how much visibility (if any) they have to the mountPath/subPath inputs the user provides. The less direct input from untrusted pvc/pod creators the drivers make use of, the less concerned I am about these aspects.

Reply:
/var/lib/kubelet/plugins/kubernetes.io/csi is where the staging mounts go, but I think we also need /var/lib/kubelet/pods/ for the per pod link right?

Regarding @liggitt's concern, I think we should disallow symlinks for the paths that get passed to the proxy.

Kubelet is supposed to give the driver only system-controlled paths, but that doesn't stop the driver from giving another path to the proxy.

Reply:
CSI drivers don't handle subpaths.  It's handled directly by the kubelet, and IIUC, this proxy is only for CSI driver file/mount operations, not kubelet.

Reply:
I'm not familiar with windows symlink mechanics. Does it have the same issue as linux where O_NOFOLLOW only applies to the past element of the path?

Reply:
I specifically mentioned for `RmdirRequest`: `Path cannot be a file of type symlink` as a restriction. The file type of the path will be checked and request to delete a path of type symlink will be rejected.

For `MkdirRequest`, the proxy creates the path and will make sure they are regular directories and not symlinks.